### PR TITLE
Annotate .ctor args in docs

### PR DESCRIPTION
**Pull request by chrsmith**
For the generated resource documentation, we were previously hard-coding the constructor arguments (in `constructor_args.tmpl`). This however isn't all that great.

- Golang needs to add a `ctx` parameter.
- The constructor args were missing their type information (i.e. https://github.com/pulumi/docs/issues/2865)

This PR fixes https://github.com/pulumi/docs/issues/2865 and tries to improve things if only slightly. It does this by:

- Adding a `ConstructorParamsTyped` (type `map[string][]formalParam`) so that we can pass along the actual set of param names and types.
- Add a new `Comment` field to `formalParam` so we can describe each ctor param.

Happy to do any refactoring or cleanup to leave the code a little better than when I found it. e.g. having the pre-rendered `ConstructorParams map[string]string` seems like a code smell. But I would guess that that's simplifying `resource.tmpl` quite a bit. (But if it's been bugging you and seems like something we'd want to merge into this PR, let me know.)

http://localhost:1313/docs/reference/pkg/kafka/acl/ **(new)**

image
image
image
image

https://www.pulumi.com/docs/reference/pkg/kafka/acl/ **(current)**

image
image

### CODE REVIEW

File: pkg/codegen/docs/gen.go [hunk 1]
```diff
@@ -469,14 +485,16 @@ func (mod *modContext) genConstructorTS(r *schema.Resource, argsOptional bool) [
...
 				Link: docLangHelper.GetDocLinkForResourceType(mod.pkg.Name, modName, argsType),
 			},
+			Comment: ctorArgsArgComment,
 		},
 		{
 			Name:         "opts",
 			OptionalFlag: "?",
 			Type: propertyType{
-				Name: "CustomResourceOptions",
+				Name: "pulumi.CustomResourceOptions",
```
Comment (praneetloke) [inline pkg/codegen/docs/gen.go:494]: Is there a reason to add the `pulumi.` back here?

**Comment by praneetloke** [comment]
> Happy to do any refactoring or cleanup to leave the code a little better than when I found it. e.g. having the pre-rendered ConstructorParams map[string]string seems like a code smell. But I would guess that that's simplifying resource.tmpl quite a bit.

Please feel free to cleanup as needed. As you said, the pre-rendering is to avoid doing some complicated things in the template as it can be hard to follow/do. Trying to mix injected variables inside markup was a little hard for these constructor params specifically due to the kind of markup it required.

File: pkg/codegen/docs/templates/constructor_args.tmpl [hunk 2]
```diff
@@ -1,19 +1,15 @@
 {{ define "constructor_args" }}
 <dl class="resources-properties">
-    <dt class="property-required" title="Required">
-        <span>name</span>
+  {{ range $params := . }}
+    <dt class="property-xxx" title="xxx">
```
Comment (praneetloke) [inline pkg/codegen/docs/templates/constructor_args.tmpl:4]: Is that class name legit? And that `title` attribute value too.

File: pkg/codegen/docs/gen.go [hunk 3]
```diff
@@ -1080,7 +1110,9 @@ func (mod *modContext) genResource(r *schema.Resource) resourceDocArgs {
 		Comment:            r.Comment,
 		DeprecationMessage: r.DeprecationMessage,
 
-		ConstructorParams:   mod.genConstructors(r, allOptionalInputs),
+		ConstructorParams:      renderedCtorParams,
+		ConstructorParamsTyped: typedCtorPArams,
```
Comment (praneetloke) [inline pkg/codegen/docs/gen.go:1114]:
```suggestion
		ConstructorParamsTyped: typedCtorParams,
```

File: pkg/codegen/docs/gen.go [hunk 4]
```diff
@@ -1071,6 +1099,8 @@ func (mod *modContext) genResource(r *schema.Resource) resourceDocArgs {
 		Notes:      mod.pkg.Attribution,
 	}
 
+	renderedCtorParams, typedCtorPArams := mod.genConstructors(r, allOptionalInputs)
```
Comment (praneetloke) [inline pkg/codegen/docs/gen.go:1102]:
```suggestion
	renderedCtorParams, typedCtorParams := mod.genConstructors(r, allOptionalInputs)
```

File: pkg/codegen/docs/gen.go [hunk 5]
```diff
@@ -567,15 +589,17 @@ func (mod *modContext) genConstructorCS(r *schema.Resource, argsOptional bool) [
...
 			},
+			Comment: ctorArgsArgComment,
 		},
 		{
 			Name:         "opts",
 			OptionalFlag: "?",
 			DefaultValue: " = null",
 			Type: propertyType{
-				Name: "CustomResourceOptions",
+				Name: "Pulumi.CustomResourceOptions",
```
Comment (praneetloke) [inline pkg/codegen/docs/gen.go:599]: Same here. Any reason to add it back?

File: pkg/codegen/docs/gen.go (hunk 1 above)
Comment (chrsmith) [reply pkg/codegen/docs/gen.go:494]: Nothing in particular, just to make it clearer that the type `CustomResourceOptions` comes from the `pulumi` package/namespace/module, rather than whatever the current resource provider is. (e.g. not `aws.CustomResourceOptions`.) But if you feel strongly about it, happy to change it back.

File: pkg/codegen/docs/templates/constructor_args.tmpl (hunk 2 above)
Comment (chrsmith) [reply pkg/codegen/docs/templates/constructor_args.tmpl:4]: It is not! Let me fix that 🙃

Comment (chrsmith) [reply pkg/codegen/docs/templates/constructor_args.tmpl]: Updated the PR to now mark params as optional/required, e.g.

image

File: pkg/codegen/docs/gen.go (hunk 1 above)
Comment (praneetloke) [reply pkg/codegen/docs/gen.go:494]: Yeah we dropped those prefixes intentionally in favor of not showing namespaces for types. FWIW, I, too, felt the same way as you did and had those prefixes in the beginning. :)

File: pkg/codegen/docs/templates/resource.tmpl [hunk 6]
```diff
@@ -28,30 +33,38 @@
...
+        <span>resource_name</span>
         <span class="property-indicator"></span>
+        <span class="property-type">str</span>
     </dt>
     <dd>The unique name of the resource.</dd>
     <dt class="property-optional" title="Optional">
         <span>opts</span>
         <span class="property-indicator"></span>
+        <span class="property-type">
+            <a href="#resource-arguments">object</a>
```
Comment (justinvp) [inline pkg/codegen/docs/templates/resource.tmpl:55]: This doesn't look right. `opts` in Python is typed as [`ResourceOptions`](https://www.pulumi.com/docs/reference/pkg/python/pulumi/#pulumi.ResourceOptions) and it should not be linking to the `resource-arguments` section, but rather to https://www.pulumi.com/docs/reference/pkg/python/pulumi/#pulumi.ResourceOptions

File: pkg/codegen/docs/gen.go (hunk 1 above)
Comment (cnunciato) [reply pkg/codegen/docs/gen.go:494]: We did decide collectively to remove these, but as I recall we didn't spend a lot of time thinking about it (I believe the intent was to conserve space), and looking at them now, they do make things clearer. I'm personally okay with putting them back in, but if others would prefer to save that kind of change for another PR, we could do that, too. A tooltip or hover (if we did want to handle it in another PR) could be a good way to strike a balance.

**Comment by chrsmith** [comment]
@justinvp thanks for the feedback, PTAL.

File: pkg/codegen/docs/gen.go [hunk 7]
```diff
@@ -535,13 +553,15 @@ func (mod *modContext) genConstructorGo(r *schema.Resource, argsOptional bool) [
 				Name: "Context",
 				Link: docLangHelper.GetDocLinkForPulumiType(mod.pkg, "Context"),
 			},
+			Comment: "Context object for the current deployment",
```
Comment (justinvp) [inline pkg/codegen/docs/gen.go:556]: @chrsmith, super nitpick... All the other comments end with a period.

Comment (justinvp) [reply pkg/codegen/docs/gen.go:556]: I had a spare moment: https://github.com/pulumi/pulumi/pull/4440
